Bring back 'Component#after_render'

jmstacey opened the pull request:
This component method callback is useful for JQuery plugins that change the real DOM and need the virtual DOM render to be patched before activating.

Example:
```ruby
def before_render
  Element['#outline'].nestable('destroy')
end

def after_render
  Element['#outline'].nestable({ ... options ...}.to_n)
end

def render
  div id: "outline"
end
```

fazibear commented:
`after_render` callback is a bad idea. They have big performance overhead. You should use virtual-dom hooks instead.

jmstacey committed:
Add 'Component#on_mounted' lifecycle callack

In the example below, on_mounted is called after the component is rendered. The will_unmount hook will fire when the component will be unmounted from the virtual DOM and before it's actually removed from the DOM.

def will_unmount
  Element['#outline'].nestable('destroy') # Deactivate plugin invalidating this outline as it should be deployed clean from the store now for one-way binding
end

def on_mounted
  Element['#outline'].nestable({ 'json': store.app.state.visible_data, 'callback': method(:on_dd_change).to_proc }.to_n) # #to_n converts to native javascript. Provided by opal-jquery
end

def render
  div class: "dd", id: "outline", hook: unhook(:will_unmount)
  . . .
end

jmstacey commented:
Renamed to ```on_mounted```. This let's us do something like the following where ```on_mounted``` callbacks are made after the component is rendered, and ```will_unmount``` when the component is being removed from the virtual DOM but before it's removed from DOM.

This would be the ideal use case, but facilitates use of legacy code that directly manipulates the DOM.

```ruby
def will_unmount
  Element['#outline'].nestable('destroy')
end

def on_mounted
  Element['#outline'].nestable({ 'json': store.app.state.visible_data, 'callback': method(:on_dd_change).to_proc }.to_n)
end

def render
  div class: "dd", id: "outline", hook: unhook(:will_unmount)
  . . .
end
```

jmstacey commented:
Reopening. Renamed the source branch and GH auto closed.

fazibear commented:
Just like I said 'after_render' callback is a very bad idea. There is performance overhead even if you don't use it. Use virtual-dom hooks instead. 

Hook gives you a reference to given node, just after mount.

 Here is the example:

```ruby
  def after_render(node)
    Element[node.to_n].nestable({ 'json': store.app.state.visible_data, 'callback': -> { on_dd_change(l, e) } }.to_n) # #to_n converts to native javascript. Provided by opal-jquery
    # Todo (caution): nestable binding is on real DOM and is likely to break on route changes. Consider a "terminate" equivalent to unload on other pages.
  end

  def render
    div class: "dd", id: "outline", hook: hook(:after_render)
    ..
  end
```
